src/sparsezoo/v2/model_directory.py

```diff
@@ -0,0 +1,502 @@
+        :param directory_path: path to the local directory
+        :return: ModelDirectory class object
+        """
+        files = []
+        paths = glob.glob(os.path.join(directory_path, "*"))
+        for path in paths:
+            display_name = os.path.basename(path)
+            files.append(file_dictionary(display_name=display_name, path=path))
+
+        return ModelDirectory(files=files, name="model_directory", path=directory_path)
+
+    def download(self, directory_path: str, override: bool = False) -> bool:
+        """
+        Attempt to download the files given the `url` attribute
+        of the files inside the ModelDirectory.
+        :param directory_path: directory to download files to
+        :param override: if True, the method can override old `directory_path`
+        :return: boolean flag; was download successful or not.
+        """
+        if self.path is not None and not override:
+            raise ValueError(
+                "ModelDirectory class object was either created "
+                "using 'from_directory` factory method or "
+                "`download()` method already invoked."
+                "Set `override` = True to override."
+            )
+        else:
+            downloads = []
+            for file in self.files:
+                downloads.append(self._download(file, directory_path))
+
+        return all(downloads)
+
+    def validate(self) -> bool:
+        """
+        Validate the ModelDirectory class object:
+        1. Validate that the sample inputs and outputs work with ONNX Runtime
+        2. Validate all the folders
+
+        return: a boolean flag; if True, the validation has been successful
+        """
+
+        if not self._validate_with_onnx_runtime():
+            logging.warning(
+                "Failed to validate the compatibility of "
+                "`sample_inputs` files with the `model.onnx` model."
+            )
+            return False
+
+        # TODO: This is a hack for now,
+        #  some files cannot be validated
+        #  using dummy inputs (see respective tests)
+        SKIP_ATTRIBUTES = ["framework_files"]
+
+        if self.path is None:
+            raise ValueError(
+                "Cannot validate the ModelDirectory. "
+                "If created using method `from_directory`, "
+                "please make sure that the `directory_path` is correct. "
+                "If created using method `from_zoo_api`, "
+                "call `download()` method prior to `validate()`"
+            )
+
+        validations = {}
+        for attribute, file in self.__iter__():
+            # TODO: Continuing with the hack
+            if attribute in SKIP_ATTRIBUTES:
+                validations[attribute] = True
+            elif isinstance(file, File):
+                validations[attribute] = file.validate()
+            elif isinstance(file, list):
+                validations[attribute] = all([_file.validate() for _file in file])
```

Comment:
```suggestion
                validations[attribute] = all(_file.validate() for _file in file)
```

This will create a generator that validates one file at a time and returns False when the first file is invalid; using a list will force it's creation entailing that `_file.validate()` is called for every file even if a previous file was invalid.
Thoughts?

Reply:
Hmm, thank you for bringing this up. I read up on the topic (e.g. https://stackoverflow.com/questions/47789/generator-expressions-vs-list-comprehensions) and I guess we can derive a heuristic:

If your computation will be super memory-heavy -> make sense to swap list comprehension to generator expression.
If you want to reuse the results of your computation -> you need to do list comprehension.

All in all @rahul-tuli while I don't think that those changes allow us to gain significant speedup, I am very grateful for those comments, definitely, I learned something there. Definitely please throw some more of those gems in the future, I enjoy learning :)

Reply:
This is correct, if you look closely here though, we are not re-using the computation, the actual list is never used again, just the result from `all(...)` but I agree this will not have a big enough impact here.